### PR TITLE
[DPE-4367] Add support for tracing through tempo-k8s

### PR DESCRIPTION
**Pull request by shayancanonical**
## Issue
We would like to be able to instrument our charm and generate traces to discover inefficiencies in charm code execution. The observability team has developed a charm for [tempo](https://github.com/canonical/tempo-k8s-operator) that we would like to integrate with.

## Solution
Integrate with the tempo charm

## Considerations
~~1. We added 2 new relation interfaces: one for `tracing` to be related with `tempo-k8s` and one `tracing-certificates` to be related with `self-signed-certificates` (or another certificates operator). The `tracing-certificates` relation is used to retrieve (with the help of the obsevability_libs `CertHandler`) the server ca cert for tempo-k8s when it has HTTPS enabled~~ We added a new relation interface for `tracing`
~~2. The tempo operator does not yet support rotation of certs, so we will need to address support for rotating certs when tempo adds this functionality~~
~~3. The `CertHandler` (by observability_libs) uses `v3` of `tls_certificates` charm lib. Our charm uses `v2`. We should upgrade to our charm to use `v3` in a follow up PR~~
4. We need to use `edge` version of the COS-lite bundle until their next stable release (as the COS team has some critical unreleased changes)
5. We are sending traces to tempo-k8s using HTTP as there are a few issues that need be addressed before we can send traces using HTTPS

image

### CODE REVIEW

**Review by paulomach — APPROVED** [review]
lgtm

File: lib/charms/tempo_k8s/v1/charm_tracing.py [hunk 1]
```diff
@@ -0,0 +1,573 @@
...
+
+def _get_server_cert(server_cert_getter, self, charm):
+    if isinstance(server_cert_getter, property):
+        server_cert = server_cert_getter.__get__(self)
+    else:  # method or callable
+        server_cert = server_cert_getter(self)
+
+    if server_cert is None:
+        logger.warning(
+            f"{charm}.{server_cert_getter} returned None; sending traces over INSECURE connection."
```
[the remark this answers is not in the document]
Comment (taurus-forever) [reply lib/charms/tempo_k8s/v1/charm_tracing.py:266]: @shayancanonical false alarm?

**Review by taurus-forever — APPROVED** [review]
Nice!

File: lib/charms/tempo_k8s/v1/charm_tracing.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (shayancanonical) [reply lib/charms/tempo_k8s/v1/charm_tracing.py:266]: yes, looks like a false alarm. it's logging the property that retrieves the server cert, not the server cert itself

**Pull request merged**